compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -996,15 +996,15 @@ object Types {
      *  (*) normalizes means: follow instantiated typevars and aliases.
      */
     def lookupRefined(name: Name)(implicit ctx: Context): Type = {
-      def loop(pre: Type): Type = pre.stripTypeVar match {
+      @tailrec def loop(pre: Type): Type = pre.stripTypeVar match {
         case pre: RefinedType =>
           pre.refinedInfo match {
             case TypeAlias(alias) =>
               if (pre.refinedName ne name) loop(pre.parent) else alias
             case _ => loop(pre.parent)
           }
         case pre: RecType =>
-          val candidate = loop(pre.parent)
+          val candidate = pre.parent.lookupRefined(name)
```

Comment:
this change is not necessary. It will still get tail-rec optimized even without it.
You can use https://github.com/lampepfl/dotty/issues/1221 to mark per-callsite `@tailrec`-s